std/assembly/util/string.ts

```diff
@@ -160,54 +180,229 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
-  // trim white spaces
-  while (isSpace(code)) {
-    code = <i32>load<u16>(ptr += 2);
+  var sign = 1.;
+  // skip white spaces
+  while (len && isSpace(code)) {
+    code = <u32>load<u16>(ptr += 2);
     --len;
   }
+  if (!len) return NaN;
+
+  // try parse '-' or '+'
   if (code == CharCode.MINUS) {
     if (!--len) return NaN;
-    code = <i32>load<u16>(ptr += 2);
-    sign = -1.0;
+    code = <u32>load<u16>(ptr += 2);
+    sign = -1;
   } else if (code == CharCode.PLUS) {
     if (!--len) return NaN;
-    code = <i32>load<u16>(ptr += 2);
+    code = <u32>load<u16>(ptr += 2);
   }
 
-  // calculate value
-  var num = 0.0;
-  while (len--) {
-    code = <i32>load<u16>(ptr);
-    if (code == CharCode.DOT) {
-      ptr += 2;
-      let fac = 0.1; // precision :(
-      while (len--) {
-        code = <i32>load<u16>(ptr);
-        if (code == CharCode.E || code == CharCode.e) {
-          assert(false); // TODO
-        }
-        code -= CharCode._0;
-        if (<u32>code > 9) break;
-        num += <f64>code * fac;
-        fac *= 0.1;
-        ptr += 2;
-      }
-      break;
+  // try parse Infinity
+  if (len == 8 && code == CharCode.I) {
+    if (
+      load<u64>(ptr, 0) == 0x690066006E0049 && // ifnI
+      load<u64>(ptr, 8) == 0x7900740069006E    // ytin
+    ) {
+      return copysign<f64>(Infinity, sign);
     }
-    code -= CharCode._0;
-    if (<u32>code >= 10) break;
-    num = num * 10 + code;
-    ptr += 2;
+    return NaN;
   }
-  return sign * num;
+  // validate next symbol
+  if (code != CharCode.DOT && <u32>(code - CharCode._0) >= 10) {
+    return NaN;
+  }
+  var savedPtr = ptr;
+  // skip zeros
+  while (code == CharCode._0) {
+    code = <u32>load<u16>(ptr += 2);
+    --len;
+  }
+  if (len <= 0) return 0;
+  const capacity = 19; // int(64 * 0.3010)
+  var pointed = false;
+  var consumed = 0;
+  var position = 0;
+  var x: u64 = 0;
+  if (code == CharCode.DOT) {
+    let noDigits = !(savedPtr - ptr);
+    ptr += 2; --len;
+    if (!len && noDigits) return NaN;
+    for (pointed = true; (code = <u32>load<u16>(ptr)) == CharCode._0; --position, ptr += 2) --len;
+    if (len <= 0) return 0;
+    if (!position && noDigits && code - CharCode._0 >= 10) return NaN;
+  }
+  for (let digit = code - CharCode._0; digit < 10 || (code == CharCode.DOT && !pointed); digit = code - CharCode._0) {
+    if (digit < 10) {
+      x = consumed < capacity ? 10 * x + digit : x | u64(!!digit);
+      ++consumed;
+    } else {
+      position = consumed;
+      pointed = true;
+    }
+    if (!--len) break;
+    code = <u32>load<u16>(ptr += 2);
+  }
+
+  if (!pointed) position = consumed;
+  return copysign<f64>(scientific(x, position - min(capacity, consumed) + parseExp(ptr, len)), sign);
+}
+
+// @ts-ignore: decorator
+@inline
+function scientific(significand: u64, exp: i32): f64 {
+  if (!significand || exp < -342) return 0;
+  if (exp > 308) return Infinity;
+  // Try use fast path
+  var result = strtodFast(<f64>significand, exp);
```

Comment:
Wondering if it would make sense to move the logic of `strtodFast` to here, saving the check for `isNaN`, doing the scaling in the `else` case.

Reply:
Good catch!